Scripted dynamic building of alpine docker image

mjwhitta opened the pull request:
I created a script to build a docker image from alpine:latest and it mirrors a user's uid and gid in the container to try to prevent any permissions issues. I thought others might find this useful, but if you feel otherwise feel free to close this request.

I've been using it for a few weeks without issue but my tests are far from exhaustive. Repeatedly running the script should be safe and will update any existing container as well as delete any old ones.

Maijin commented:
What's the added value with https://github.com/radare/radare2/blob/master/Dockerfile which is already been used and tested via Travis-CI, what about having this one in extras?

Maijin commented:
For the check you can use https://github.com/radare/radare2/blob/master/sys/shellcheck.sh

mjwhitta commented:
The added value here is twofold. First, the size is only 140MB instead of 1GB. Second, files created in this Docker container should have the same permissions/ownership as the user when they exit the container, as opposed to `root:root` with the old container.

I'm comfortable moving this to sys. The `r2-docker.sh` script is there so it makes some sense (enough that I probably should have already put it there). I'm unsure what extras is as I don't see that directory.

The shellcheck.sh script threw some errors as I was missing dependencies. Should I fix that too? Regardless, I addressed most of the shellsheck issues other than complaints about literal backslashes as those look like false positives to me. Is it preferred that I add a new commit or amend my current one?

ret2libc commented:
> What's the added value with https://github.com/radare/radare2/blob/master/Dockerfile which is already been used and tested via Travis-CI, what about having this one in extras?

@Maijin we use Dockerfile.travis actually AFAIK.

Maijin commented:
ok. For the shellcheck and the change of directory you can just push -f your changes on the PR will work

mjwhitta commented:
Ok, I think I made all the necessary changes. I didn't fix the complaints about literal backslashes b/c I found [this](https://github.com/koalaman/shellcheck/wiki/SC1117#retired).

radare commented on sys/docker_build_alpine_image:
```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env bash
```
use /bin/sh, no bashisms allowed in r2land

radare replied:
add .sh to the filename

mjwhitta replied:
Changing to /bin/sh will likely mean a rewrite. I had to make some small changes but shellcheck now complains about lack of array support. It still runs but this could be because /bin/sh is a symlink to /bin/bash on my box. I thought that would cause bash to run in a more /bin/sh manner but I'm not sure how consistent that is across platforms/OS.

radare commented on sys/docker_build_alpine_image:
```diff
@@ -0,0 +1,328 @@
+# Install bash b/c it's better
+RUN apk upgrade && apk add bash
+
+# Bash is better than sh
```
that depends on your needs/taste

radare commented on sys/docker_build_alpine_image:
```diff
@@ -0,0 +1,328 @@
+It's suggested you add something like the following to your ~/.bashrc:
+
+r2d() {
+    local rm
```
i prefer variablenames in uppercase in shellscripts

mjwhitta replied:
I thought uppercase variables were for variables that get exported or inherited from the env. Is that not a standard? I'm pretty sure shellcheck complains about that.

radare commented on sys/docker_build_alpine_image:
```diff
@@ -0,0 +1,328 @@
+            docker images | awk "/$old_base/ {print \$3}"
+        ); unset id
+    fi
+fi; unset found
```
move this unset below

mjwhitta replied:
I moved all `unset` lines to a newline.

radare commented on sys/docker_build_alpine_image:
```diff
@@ -0,0 +1,328 @@
+[[ -z $help ]] || usage 0
+[[ $# -eq 0 ]] || usage 1
+if [[ $(id -u) -ne 0 ]] && id -Gn | grep -qvw "docker"; then
+    errx 2 "You are not part of the docker group"
```
use tabs

mjwhitta replied:
Done.

mjwhitta commented:
Do arrays count as bashisms? `shellcheck` complains about arrays and the use of `[[ ]]` now, but it's 2018, I thought `[[ ]]` was preferred over `[ ]` for boolean tests like if statements and while loops. I don't work with `/bin/sh` often so I'm not aware of its limitations.

radare commented:
For checks you can just use `[ ]`. about arrays, yep, thats a bash specific feature. use space separated strings for example

mjwhitta commented:
Ok, I think I removed all bashisms mentioned by `shellcheck` after I changed the header. As far as I can tell it works without issue. Let me know if there's anything else.

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+}
+long_opt() {
+	arg=""
+    shift="0"
```
tabs

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+		if [ -z "$(command -v "$d")" ]; then
+            echo "[!] $d is not installed"
+            exit 128
+        fi
```
tabs

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+[ $# -eq 0 ] || usage 1
+if [ "$(id -u)" -ne 0 ] && id -Gn | grep -qvw "docker"; then
+	echo "[!] You are not part of the docker group"
+    exit 2
```
tabs

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+echo "[*] Tagging any old r2_alpine images"
+docker images | awk '{print $1":"$3}' | while read -r tag; do
+	case "$tag" in
+		"r2_alpine:"*) docker image tag "${tag#*:}" "$tag" ;;
```
no need to indent the cases

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+	echo "[*] Old images:"
+	docker images | head -n 1
+
+    docker images | grep -E "($old_base)|($old_r2)" | \
```
tabs

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+	echo
+	while :; do
+		echo "Remove old images (y/N/q)?"
+        read -r ans
```
tabs

radare commented on sys/docker_build_alpine_image.sh:
```diff
@@ -0,0 +1,328 @@
+        read -r ans
+		echo
+		case "$ans" in
+			""|"n"|"N"|"q"|"Q") break ;;
```
no need to indent the cases

mjwhitta commented:
Oops! I thought I got all those spaces, will fix.

radare commented:
Thanks!
